src/Service/MemoryInfoHandler.cpp

```diff
@@ -49,26 +54,64 @@ void MemoryInfoHandler::Stop() {
 
 void MemoryInfoHandler::OnCGroupMemoryUsage(
     orbit_grpc_protos::CGroupMemoryUsage cgroup_memory_usage) {
-  orbit_grpc_protos::ProducerCaptureEvent event;
-  *event.mutable_cgroup_memory_usage() = std::move(cgroup_memory_usage);
-  producer_event_processor_->ProcessEvent(orbit_grpc_protos::kMemoryInfoProducerId,
-                                          std::move(event));
+  uint64_t sampling_window_id = GetSamplingWindowId(cgroup_memory_usage.timestamp_ns());
+
+  absl::MutexLock lock(&in_progress_wrappers_mutex_);
+  *in_progress_wrappers_[sampling_window_id].mutable_cgroup_memory_usage() =
+      std::move(cgroup_memory_usage);
+  ProcessMemoryEventWrapperIfReady(sampling_window_id);
 }
 
 void MemoryInfoHandler::OnProcessMemoryUsage(
     orbit_grpc_protos::ProcessMemoryUsage process_memory_usage) {
-  orbit_grpc_protos::ProducerCaptureEvent event;
-  *event.mutable_process_memory_usage() = std::move(process_memory_usage);
-  producer_event_processor_->ProcessEvent(orbit_grpc_protos::kMemoryInfoProducerId,
-                                          std::move(event));
+  uint64_t sampling_window_id = GetSamplingWindowId(process_memory_usage.timestamp_ns());
+
+  absl::MutexLock lock(&in_progress_wrappers_mutex_);
+  *in_progress_wrappers_[sampling_window_id].mutable_process_memory_usage() =
+      std::move(process_memory_usage);
+  ProcessMemoryEventWrapperIfReady(sampling_window_id);
 }
 
 void MemoryInfoHandler::OnSystemMemoryUsage(
     orbit_grpc_protos::SystemMemoryUsage system_memory_usage) {
+  uint64_t sampling_window_id = GetSamplingWindowId(system_memory_usage.timestamp_ns());
+
+  absl::MutexLock lock(&in_progress_wrappers_mutex_);
+  *in_progress_wrappers_[sampling_window_id].mutable_system_memory_usage() =
+      std::move(system_memory_usage);
+  ProcessMemoryEventWrapperIfReady(sampling_window_id);
+}
+
+uint64_t MemoryInfoHandler::GetSamplingWindowId(uint64_t sample_timestamp_ns) const {
+  return static_cast<uint64_t>(
+      std::round(static_cast<double>(sample_timestamp_ns - sampling_start_timestamp_ns_) /
+                 sampling_period_ns_));
+}
+
+void MemoryInfoHandler::ProcessMemoryEventWrapperIfReady(uint64_t sampling_window_id) {
+  if (!in_progress_wrappers_.contains(sampling_window_id)) return;
+
+  orbit_grpc_protos::MemoryEventWrapper& wrapper = in_progress_wrappers_[sampling_window_id];
+  bool wrapper_is_ready = wrapper.has_system_memory_usage();
+  if (enable_cgroup_memory_) {
+    wrapper_is_ready =
+        wrapper_is_ready && wrapper.has_process_memory_usage() && wrapper.has_cgroup_memory_usage();
+  }
+  if (!wrapper_is_ready) return;
+
+  uint64_t synchronized_timestamp_ns = wrapper.system_memory_usage().timestamp_ns();
+  if (enable_cgroup_memory_) {
+    synchronized_timestamp_ns = static_cast<uint64_t>(
+        (synchronized_timestamp_ns + wrapper.cgroup_memory_usage().timestamp_ns() +
+         wrapper.process_memory_usage().timestamp_ns()) /
+        3.0);
```

Comment:
This deserves a comment explaining why you went with the artihmetic mean and not with any other averaging function.

Reply:
Also, why `3.0`? This means it will be a floating point division and then immediately casted back to uint64_t. Can you explain the advantage of making a float point division here?

Reply:
Also we're getting pretty close to the maximum value of `uint64_t` when adding up 3 nanosecond timestamps.

This is fine for now, but if someone wants to add more timestamps to the averaging in the future this might overflow. I suggest to add a comment and to warn future editors of this code.

Reply:
Does someone know how timestamp_ns is measuring? I've just tested 2 different captures with 2 days and an hour of difference and the timestamps seems to be 6.29 days and 8.33 days. Unless we arrive to 500 years, we shouldn't have any issue.

But we can add a comment to assure future editors we won't have a problem there.

Reply:
So it's not a unix timestamp?

Reply:
It's a monotonic timestamp, nanoseconds since boot.

Reply:
BTW, I still think than using average could confuse users. Shouldn't we use instead min or max timestamp to combine them?

Reply:
> It's a monotonic timestamp, nanoseconds since boot.
Then it shouldn't be a problem.

> BTW, I still think than using average could confuse users. Shouldn't we use instead min or max timestamp to combine them?
That sounds even more confusing to me.

Reply:
ok

Reply:
Thanks for the suggestion. This part has been changed accordingly and moved to the `MemoryInfoListener::GetSynchronizedSamplingTimestamp` method which returns the arithmetic mean.